extra/DEVELOPER_DOCS/Language.md

```diff
@@ -0,0 +1,136 @@
+
+### Creating a pipeline component from a factory
+
+`Language.add_pipe` takes care of adding a pipeline component, given its factory name, its config. If no source pipeline to copy the component from is provided, it delegates to `Language.create_pipe`, which sets up the actual component function.
+
+- Validate the config and make sure that the factory was registered via the decorator and that we have meta for it.
+- Update the component config with any defaults specified by the component's `default_config`, if available. This is done by merging the values we receive into the defaults. It ensures that you can still add a component without having to specify its _entire_ config including more complex settings like `model`. If no `model` is defined, we use the default.
+- Check if we have a language-specific factory for the given `nlp.lang` and if not, fall back to the global factory.
+- Construct the component config, consisting of whatever arguments were provided, plus the current `nlp` object and `name`, which are default expected arguments of all factories. We also add a reference to the `@factories` registry, so we can resolve the config via the registry, like any other config. With the added `nlp` and `name`, it should now include all expected arguments of the given function.
+- Fill the config to make sure all unspecified defaults from the function arguments are added and update the `raw_config` (uniterpolated with variables intact) with that information, so the component config we store in `nlp.config` is up to date. We do this by adding the `raw_config` _into_ the filled config – otherwise, the references to variables would be overwritten.
+- Resolve the config and create all functions it refers to (e.g. `model`). This gives us the actual component function that we can insert into the pipeline.
+
+### Sourcing a pipeline component
+
+spaCy also allows "sourcing" a component, which will copy it over from an existing pipeline. In this case, `Language.add_pipe` will delegate to `Language.create_pipe_from_source`. In order to copy a component effectively and validate it, the source pipeline first needs to be loaded. This is done in `Language.from_config`, so a source pipeline only has to be loaded once if multiple components source from it. Sourcing a component will perform the following checks and modifications:
+
+- If the source pipeline contains vectors, check that they match the destination pipeline's vectors and warn if not (since different vectors that are used as features in components can lead to degraded performance). This is done by comparing the shape first and finally falls back to comparing the actual byte representation of the vectors (which is slower). We also store a hash of the vectors in the pipeline meta so we're able to check this more easily.
```

Comment:
I think this is missing an explanation of the delayed vector loading so that they're not loaded twice. The only reason for storing the hash is for `source` from a config, not for `create_pipe_from_source` itself.

Reply:
Good point! Do you want to add that part? Since you implemented the source hashing, I think you might be able to explain this better than I can.